File: day11/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>
#include <sstream>
#include <unordered_set>
#include <unordered_map>
#include <map>
#include <list>
#include <regex>

auto read() {
    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");
    std::string line;
    std::vector<std::string> result;
    while (std::getline(ifs, line)) {
        if (line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}

auto read_all() {
    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");
    std::stringstream data;
    data << ifs.rdbuf();
    return data.str();
}

std::vector<std::string> split(const std::string &str, char c = ' ') {
    std::istringstream split(str);
    std::vector<std::string> tokens;
    for (std::string each; std::getline(split, each, c); tokens.push_back(each));
    return tokens;
}

struct Monkey {
    std::deque<int64_t> items;
    std::pair<char, std::string> op;
    int64_t test_div = 0;
    int64_t test_true = -1;
    int64_t test_false = -1;
    int64_t inspection_times = 0;
};

std::vector<Monkey> parse() {
    std::regex words_regex("Monkey ([0-9]+):\n"
                           "  Starting items: ([0-9, ]+)\n"
                           "  Operation: new = old ([*+]) ([0-9]+|old)\n"
                           "  Test: divisible by ([0-9]+)\n"
                           "    If true: throw to monkey ([0-9]+)\n"
                           "    If false: throw to monkey ([0-9]+)", std::regex::extended);

    auto data = read_all();

    std::vector<Monkey> monkeys;
    for (std::sregex_iterator it = std::sregex_iterator(data.begin(), data.end(), words_regex);
         it != std::sregex_iterator(); ++it) {
        monkeys.push_back({});
        std::smatch match = *it;
        auto items = split(match[2].str(), ',');
        std::transform(items.begin(), items.end(), std::back_inserter(monkeys.back().items),
                       [](std::string str) { return std::stoi(str); });
        monkeys.back().op = {match[3].str()[0], match[4].str()};
        monkeys.back().test_div = std::stoi(match[5]);
        monkeys.back().test_true = std::stoi(match[6]);
        monkeys.back().test_false = std::stoi(match[7]);
    }
    return monkeys;
}

void part1() {

    auto monkeys = parse();
    for (int i = 0; i < 20; ++i) {
        for (int j = 0; j < monkeys.size(); ++j) {
            auto &monkey = monkeys[j];
            monkey.inspection_times += monkey.items.size();
            for (auto it = monkey.items.begin(); it != monkey.items.end();) {
                auto value = *it;
                it = monkeys[j].items.erase(it);

                int64_t right_hand = monkey.op.second == "old" ? value : std::stoi(monkey.op.second);
                if (monkey.op.first == '+')
                    value += right_hand;
                else if (monkey.op.first == '*')
                    value *= right_hand;
                else
                    std::abort();

                value /= 3;
                if (value % monkey.test_div == 0)
                    monkeys[monkey.test_true].items.push_back(value);
                else
                    monkeys[monkey.test_false].items.push_back(value);
            }
        }
    }

    std::vector<int64_t> inspection_times;
    for (const auto &monkey: monkeys)
        inspection_times.push_back(monkey.inspection_times);

    std::sort(inspection_times.begin(), inspection_times.end(), std::greater<int64_t>());
    std::cout << "Part1: " << inspection_times[0] * inspection_times[1] << std::endl;
}

void part2() {

    auto monkeys = parse();

    int64_t master_modulo = 1;
    for (const auto &monkey: monkeys) {
        master_modulo *= monkey.test_div;
    }
    for (int i = 0; i < 10000; ++i) {
        for (auto &monkey: monkeys) {
            monkey.inspection_times += monkey.items.size();
            for (auto it = monkey.items.begin(); it != monkey.items.end();) {
                auto value = *it % master_modulo + master_modulo;
                it = monkey.items.erase(it);

                auto right_hand = monkey.op.second == "old" ? value : std::stoi(monkey.op.second);
                if (monkey.op.first == '+') {
                    assert(std::numeric_limits<int64_t>::max() - right_hand > value);
                    value += right_hand;
                } else if (monkey.op.first == '*') {
                    assert(std::numeric_limits<int64_t>::max() / right_hand > value);
                    value *= right_hand;
                } else
                    std::abort();

                if (value % monkey.test_div == 0)
                    monkeys[monkey.test_true].items.push_back(value);
                else
                    monkeys[monkey.test_false].items.push_back(value);
            }
        }
    }

    std::vector<int64_t> inspection_times;
    for (const auto &monkey: monkeys)
        inspection_times.push_back(monkey.inspection_times);

    std::sort(inspection_times.begin(), inspection_times.end(), std::greater<int64_t>());
    std::cout << "Part2: " << inspection_times[0] * inspection_times[1] << std::endl;
}

int main() {
    part1();
    part2();
    return 0;
}

File: day11/src/main.rs
use std::cell::RefCell;
use std::cmp::Ordering;
use std::collections::VecDeque;
use std::fs::File;
use std::io::{self, BufRead};
use std::ops::Deref;
use std::path::Path;
use std::rc::Rc;
use std::vec;
use regex::Regex;
use scanf::sscanf;
use crate::Value::{IntVal, OldVal};
use lazy_static::lazy_static;

fn read_all() -> String {
    let file_path = "input.txt";
    std::fs::read_to_string(file_path).unwrap()
}

fn read_lines() -> Vec<String> {
    read_all().split('\n').map(|x| x.parse().unwrap()).collect::<Vec<String>>()
}

enum Value {
    IntVal(i64),
    OldVal,
}

struct Monkey {
    items: Vec<i64>,
    op: (String, Value),
    test_div: i64,
    test_true: i64,
    test_false: i64,
    inspection_times: usize,
}

impl Monkey {
    fn new() -> Monkey {
        Monkey {
            items: Default::default(),
            op: ("+".parse().unwrap(), OldVal),
            test_div: 0,
            test_true: 0,
            test_false: 0,
            inspection_times: 0,
        }
    }
}

fn parse() -> Vec<Monkey> {
     lazy_static! {
        static ref RE: Regex = Regex::new(
r#"Monkey ([0-9]+):
\s*Starting items: (?P<items>[0-9, ]+)
\s*Operation: new = old (?P<op>[*+]) (?P<value>[0-9]+|old)
\s*Test: divisible by (?P<div>[0-9]+)
\s*If true: throw to monkey (?P<true>[0-9]+)
\s*If false: throw to monkey (?P<false>[0-9]+)"#).unwrap();
    }

    let text = read_all();

    let mut monkeys: Vec<Monkey> = vec![];
    for cap in RE.captures_iter(&*text) {
        let mut monkey = Monkey::new();
        monkey.items = cap.name("items").unwrap().split(',').map(|x| x.trim().parse::<i64>().unwrap()).collect::<Vec<i64>>();
        match cap.name("value") {
            Some("old") => {
                monkey.op = (cap[3].parse().unwrap(), OldVal);
            }
            Some(intval) => {
                monkey.op = (cap[3].parse().unwrap(), IntVal(intval.parse::<i64>().unwrap()));
            }
            _ => panic!("Unknown operation")
        }

        monkey.test_div = cap.name("div").unwrap().parse::<i64>().unwrap();
        monkey.test_true = cap.name("true").unwrap().parse::<i64>().unwrap();
        monkey.test_false = cap.name("false").unwrap().parse::<i64>().unwrap();

        monkeys.push(monkey);
    }
    monkeys
}

fn part1() {
    let mut monkeys = parse();
    for _ in 0..20 {
        for idx in 0..monkeys.len() {
            monkeys[idx].inspection_times += monkeys[idx].items.len();
            let items = monkeys[idx].items.drain(..).collect::<Vec<i64>>();
            for mut item in items {
                match (monkeys[idx].op.0.as_str(), &monkeys[idx].op.1) {
                    ("*", OldVal) => {
                        item *= item;
                    }
                    ("+", OldVal) => {
                        item += item;
                    }
                    ("*", IntVal(x)) => {
                        item *= x;
                    }
                    ("+", IntVal(x)) => {
                        item += x;
                    }
                    (_, _) => panic!("Invalid operation")
                }

                item /= 3;

                let dest_idx;
                if item % monkeys[idx].test_div == 0 {
                    dest_idx = monkeys[idx].test_true;
                } else {
                    dest_idx = monkeys[idx].test_false;
                }
                monkeys[dest_idx as usize].items.push(item);
            }
        }
    }

    let mut times = monkeys.iter().map(|x| x.inspection_times).collect::<Vec<usize>>();
    times.sort_by(|a, b| b.cmp(a));

    let result = times[0..2].iter().cloned().reduce(|accum, x| accum * x).unwrap();
    println!("Part1: {}", result)
}

fn part2() {
    let mut monkeys = parse();

    let master_mod = monkeys.iter().map(|x| x.test_div).reduce(|accum, x| accum * x).unwrap();

    for _ in 0..10000 {
        for idx in 0..monkeys.len() {
            monkeys[idx].inspection_times += monkeys[idx].items.len();
            let items = monkeys[idx].items.drain(..).collect::<Vec<i64>>();
            for mut item in items {
                item = item % master_mod;
                match (monkeys[idx].op.0.as_str(), &monkeys[idx].op.1) {
                    ("*", OldVal) => {
                        item *= item;
                    }
                    ("+", OldVal) => {
                        item += item;
                    }
                    ("*", IntVal(x)) => {
                        item *= x;
                    }
                    ("+", IntVal(x)) => {
                        item += x;
                    }
                    (_, _) => panic!("Invalid operation")
                }

                let dest_idx;
                if item % monkeys[idx].test_div == 0 {
                    dest_idx = monkeys[idx].test_true;
                } else {
                    dest_idx = monkeys[idx].test_false;
                }
                monkeys[dest_idx as usize].items.push(item);
            }
        }
    }

    let mut times = monkeys.iter().map(|x| x.inspection_times).collect::<Vec<usize>>();
    times.sort_by(|a, b| b.cmp(a));

    let result = times[0..2].iter().cloned().reduce(|accum, x| accum * x).unwrap();
    println!("Part2: {}", result);
}


fn main() {
    part1();
    part2();
}


